app/services/bulk_create_test_applications.rb

```diff
@@ -0,0 +1,89 @@
+class BulkCreateTestApplications
+  attr_reader :template_application_form, :candidate, :application_choice
+
+  def initialize(template_application_form)
+    @template_application_form = template_application_form
+    @candidate = template_application_form.candidate
+    @application_choice = template_application_form.application_choices.first
+  end
+
+  def call
+    error_if_application_choice_state_is_not_unsubmitted
+
+    sql = compose_insert_statements(unique_id: SecureRandom.hex(10))
+    begin
+      tries ||= 0
+      ActiveRecord::Base.connection.execute(sql)
+    rescue ActiveRecord::RecordNotUnique
+      retry unless (tries += 1) > 3
```

Comment:
Does the `sql =` declaration need to be in the begin block, otherwise `retry` will just try running the same unique_id?

Reply:
:facepalm: it does! Fixed.